design/service-apis-implementation.md

```diff
@@ -0,0 +1,229 @@
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
+
+### Status management
+
+Each object in the Service APIs set has its own status definition. In general, Contour will update the status of any object that comes into scope with its details, ensuring that the `observedGeneration` field is set correctly.
+When objects fall out of scope, any status set by Contour will not be removed.
+It's expected that things that check the status will also check the `observedGeneration` to check if the status information is up-to-date with the current object generation.
+
+### Per-object design notes
+
+#### GatewayClass
+
+Contour will look for entries with `spec.controller` equal to `projectcontour.io/contour` by default.
+Contour can change this value to any `projectcontour.io/<value>` with a config setting (similar to how `ingress.class` works today.)
+Importantly, this means that changing this values requires a Contour restart; this means that the GatewayClass watcher can have this value baked in at runtime.
+
+If Contour finds more than one matching GatewayClass, this is a conflict, and Contour will use the oldest GatewayClass (by creation timestamp).
+All GatewayClasses will be updated with status conditions indicating if they are in use.
+
+The output of this watcher is a list of GatewayClass name/namespace details.
+
+#### Gateway
+Contour will watch for entries with known GatewayClass name/namespace details, from the previous watch.
+
+Contour is not able to specify the address that the Envoy deployment will listen on, and so having any entries in Gateway's `spec.addresses` stanza will result in Contour marking those addresses as invalid in the Gateway's status.
+Contour expects the `spec.addresses` field to be empty.
+
+For Listeners:
+- Listeners inside Gateways will be merged where possible, as will listeners across Gateways.
+- Conflicts within a Gateway will result in the relevant Listeners both being rejected.
+- Conflicts between Gateways will result in the oldest Gateway (by creation timestamp) being kept.
+- The rejected Listener will have its status updated with the reason and the name of the conflicting Gateway.
+- Listeners are considered conflicting if all the fields out of `hostname`, `port`, `protocol`, and `tls` match.
```

Comment:
Could TLS listeners also be in conflict if they reference different certificates?

Reply:
relevant: https://github.com/projectcontour/contour/pull/3213/files#r552799600

Reply:
I think I've covered this in a recent update, please tell me if it's clearer now.